### PR TITLE
ffi: Add class for key-value pair log events.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

This PR introduces a new class `KeyValuePairLogEvent`. This class represents a key-value pair log event serialized by the key-value pair IR format. It consists of the following members:
- An unordered map of key-value pairs. Each key-value pair consists of an id (as the key) and a `Value`, where the id refers to the schema tree node id.
- A shared pointer to the schema tree.
- An UTC offset.

An instance must be created using the factory function. The factory function will validate whether the value's type matches the corresponding key. The input of the factory function should come from the deserialization methods, which will be used in the coming PRs.

# Validation performed

- Ensure the class can be successfully built.
- Tested in development branches that this class is functional with key-value pair IR format deserializer.
- Add unit tests to cover errors on creation.

### CODE REVIEW

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include "SchemaTree.hpp"
+#include "SchemaTreeNode.hpp"
+#include "Value.hpp"
+
+namespace clp::ffi {
+/**
+ * Class for key-value pair log events. Each key-value pair log event contains the following items:
+ *  - A list of key-value pairs
+ *  - A reference to the schema tree
+ *  - The UTC offset of the current log event
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:21]:
```suggestion
 * A log event containing key-value pairs. Each event contains:
 *  - A collection of node-id & value pairs, where each pair represents a leaf `SchemaTreeNode` in
 *    the `SchemaTree`.
 *   `SchemaTree` and the key is the node's ID.
 * - A reference to the `SchemaTree`
 * - The UTC offset of the current log event
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ */
+class KeyValuePairLogEvent {
+public:
+    // Types
+    using KeyValuePairs = std::unordered_map<SchemaTreeNode::id_t, std::optional<Value>>;
+
+    // Factory functions
+    /**
+     * Creates a key-value pair log event from valid given inputs.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:30]:
```suggestion
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+     * @param schema_tree
+     * @param kv_pairs
+     * @param utc_offset
+     * @return A result containing the key-value pair log event or an error code indicating the
+     * failure:
+     * - std::errc::operation_not_permitted if the key ID doesn't represent a valid node in the
+     *   schema tree.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:37]:
```suggestion
     * - std::errc::operation_not_permitted if a key ID doesn't represent a valid node in the schema
     *   tree.
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+using std::string;
+
+namespace clp::ffi {
+namespace {
+/**
+ * @param type
+ * @param value
+ * @return Whether the given schema tree node type matches the given value's type.
+ */
+[[nodiscard]] auto is_valid_value_type(SchemaTreeNode::Type type, Value const& value) -> bool;
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:29]: How about `node_type_matches_value_type`?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                    // The key is duplicated
+                    return std::errc::protocol_not_supported;
+                }
+            } else {
+                key_sets.emplace(parent_id, std::unordered_set{string{key_name}});
+            }
+        }
+    } catch (SchemaTree::OperationFailed const& ex) {
+        return std::errc::operation_not_permitted;
+    }
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:88]: Can we move this into a new function to reduce the complexity of this factory?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+}
+}  // namespace
+
+auto KeyValuePairLogEvent::create(
+        std::shared_ptr<SchemaTree> schema_tree,
+        KeyValuePairs kv_pairs,
+        UtcOffset utc_offset
+) -> OUTCOME_V2_NAMESPACE::std_result<KeyValuePairLogEvent> {
+    std::unordered_map<SchemaTreeNode::id_t, std::unordered_set<string>> key_sets;
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:57]: * Since the schema tree (and by extension, the key names) will continue to exist for the duration of this method, can we use a `string_view`?
* Instead of `key_sets`, how about `parent_node_id_to_key_names`?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    try {
+        for (auto const& [key_id, value] : kv_pairs) {
+            if (SchemaTree::cRootId == key_id) {
+                return std::errc::protocol_error;
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:61]: Perhaps this should be `operation_not_permitted` since the caller isn't permitted to pass in a non-leaf node ID?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            auto const& node{schema_tree->get_node(key_id)};
+            auto const type{node.get_type()};
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:65]: How about `node_type`? `type` is a bit too ambiguous.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            if (false == value.has_value()) {
+                // Empty value
+                if (SchemaTreeNode::Type::Obj != type) {
+                    return std::errc::protocol_error;
+                }
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:70]: I kinda feel like this can be moved into the value type validation method.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    try {
+        for (auto const& [key_id, value] : kv_pairs) {
+            if (SchemaTree::cRootId == key_id) {
+                return std::errc::protocol_error;
+            }
+
+            auto const& node{schema_tree->get_node(key_id)};
+            auto const type{node.get_type()};
+            if (false == value.has_value()) {
+                // Empty value
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:67]:
```suggestion
                // Value is an empty object (`{}`, which is not the same as `null`)
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                if (SchemaTreeNode::Type::Obj != type) {
+                    return std::errc::protocol_error;
+                }
+            } else if (false == is_valid_value_type(type, value.value())) {
+                return std::errc::protocol_error;
+            }
+
+            auto const parent_id{node.get_parent_id()};
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:75]: How about `parent_node_id`?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Class for key-value pair log events. Each key-value pair log event contains the following items:
+ *  - A list of key-value pairs
+ *  - A reference to the schema tree
+ *  - The UTC offset of the current log event
+ */
+class KeyValuePairLogEvent {
+public:
+    // Types
+    using KeyValuePairs = std::unordered_map<SchemaTreeNode::id_t, std::optional<Value>>;
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:26]: How about `NodeIdValuePairs`? I think `KeyValuePairLogEvent` is an appropriate name since it represents the abstraction, but for the more concrete types and variables, I think it's clearer to name them more directly.

Similarly, I would rename the `kv_pair` variables to `node_id_value_pairs`.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        case SchemaTreeNode::Type::Str:
+            return value.is<string>() || value.is<FourByteEncodedTextAst>()
+                   || value.is<EightByteEncodedTextAst>();
+        default:
+            return false;
+    }
+}
+}  // namespace
+
+auto KeyValuePairLogEvent::create(
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:52]: Can we write a simple unit test to test the error returns? I think you could re-use the schema tree you built in the schema tree test and then it should(?) be easy to create some kv-pairs with errors. (And you should naturally be able to write a test without errors.)

I know you'll eventually test this class with the deserializer, but the focus of those tests probably be to serialize and test correct deserialization rather than testing corrupt data, so that's why I think the test suggested above is valuable.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 9 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp]: prefer to leave it outside for two reasons:
1. empty isn't a `Value` type
2. This simplifies the helper function so it doesn't have to check `has_value()` and we don't need to symbols inside the helper (one for the `std::optional<Value>` and the over for the reference of the underlying `Value`)

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+/**
+ * @param schema_tree
+ * @param node_id
+ * @param node_id_value_pairs
+ * @return Whether the given node is a leaf node in the sub schema tree defined by
+ * `node_id_value_pairs`. A node is considered a leaf if none of its descendants appear in the
+ * `node_id_value_pairs`.
+ */
+[[nodiscard]] auto is_leaf_node(
```
Comment (haiqi96) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:59]: Disclaimer: I only have a rough understanding on how schema tree works.
This function is bit confusing for me. 
"A node is considered a leaf if none of its descendants appear in the `node_id_value_pairs`." <- This doesn't seem to be the normal definition for a leaf node. Is this expected?

In addition, "the sub schema tree defined by `node_id_value_pairs`." Is there any guarantee that nodes in the `node_id_value_pairs` will form a subtree? Just looking at the argument without thinking about how it is generated, it is possible node_id_value_pairs contains a list of unconnected nodes.

And lastly a highlevel comment is that I feel this function should belong to the Schema tree class.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        for (auto const& [node_id, value] : node_id_value_pairs) {
+            if (SchemaTree::cRootId == node_id) {
+                ret_val.emplace(std::errc::operation_not_permitted);
+                break;
+            }
+
+            auto const& node{schema_tree.get_node(node_id)};
+            auto const node_type{node.get_type()};
+            if (false == value.has_value()) {
+                // Value is an empty object (`{}`, which is not the same as `null`)
```
Comment (haiqi96) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:102]: I assume this comment tries to explain why the line doesn't use `value.is_null();` in a way similar to line 68?
I don't think I follow the comment though. Can you clarify a bit?

And we can combine the two if statements?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            std::shared_ptr<SchemaTree> schema_tree,
+            NodeIdValuePairs node_id_value_pairs,
+            UtcOffset utc_offset
+    )
+            : m_schema_tree{std::move(schema_tree)},
+              m_node_id_value_pairs{std::move(node_id_value_pairs)},
+              m_utc_offset{utc_offset} {}
+
+    // Variables
+    std::shared_ptr<SchemaTree> m_schema_tree;
```
Comment (haiqi96) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:81]: Techinically it is possible to modify the original schema tree using this pointer in the class, right?
Should we consider using std::shared_ptr<SchemaTree const> ?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    // Disable copy constructor and assignment operator
+    KeyValuePairLogEvent(KeyValuePairLogEvent const&) = delete;
+    auto operator=(KeyValuePairLogEvent const&) -> KeyValuePairLogEvent& = delete;
+
+    // Default move constructor and assignment operator
+    KeyValuePairLogEvent(KeyValuePairLogEvent&&) = default;
+    auto operator=(KeyValuePairLogEvent&&) -> KeyValuePairLogEvent& = default;
+
+    // Destructor
+    ~KeyValuePairLogEvent() = default;
```
Comment (haiqi96) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:58]: I believe we don't need to define the destructor if it's default?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+
+    [[nodiscard]] auto get_node_id_value_pairs() const -> NodeIdValuePairs const& {
+        return m_node_id_value_pairs;
+    }
+
+    [[nodiscard]] auto get_utc_offset() const -> UtcOffset { return m_utc_offset; }
+
+private:
+    // Constructor
+    KeyValuePairLogEvent(
```
Comment (haiqi96) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:71]: do we need explicit?

**Review by haiqi96 — COMMENTED** [review]
Just quickly skimmed through the class and haven't read the unit tests yet.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 15 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:102]: 1. `{}` and `null` are two different types of values.
2. I'd prefer to handle them separately. Combining these two statements will make it very hard to read

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp (hunk 16 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:81]: Make sense

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp (hunk 17 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:58]: To silent `clang-tidy` warnings, lol

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 15 above)
Comment (haiqi96) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:102]: I still don't get when there will be {} and when there will be null (and why we only check for {} but not null in this function)
Is it because null is already handled in the validation? If so, I think you should leave this information in the comment.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp (hunk 17 above)
Comment (haiqi96) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:58]: Uh, I remember I defined something like this and got a comment asked me to remove it (or maybe it was a constructor). maybe we should really agree on the standard for a default constructor and destructor

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 14 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:59]: This is a validation method: if an object type node appears in the key value pairs, it must be a leaf node. The way we store key value pairs is to store the schema (all as leaf nodes), so the path from the root to the leaves are implicitly indicated. In the merged schema tree, usually an object type node is a non leaf node. But in a key value pairs' schema, this node can be a leaf node as long as its value is `{}` or `null`. Our goal is to validate that if such a node appears, all its descendants in the merged schema tree can't appear in the schema of the given key value pair log event. `it is possible node_id_value_pairs contains a list of unconnected nodes` isn't true since everything will be connected to the root if they are valid leaf nodes. This function is used to validate pairs using schema tree's information. I think we should split it from a schema tree's method.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 15 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:102]: `{}` is considered as a special case, where `null` is a valid primitive value (like an int or a string). These are user-input values.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp (hunk 18 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:71]: Yes

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp (hunk 17 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:58]: The warning is that if u defined either copy or move constructor, u will have to explicitly define the destructor (even declare it to be the default)

Comment (haiqi96) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:58]: Fair enough

Comment (kirkrodrigues) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:58]: http://scottmeyers.blogspot.com/2014/03/a-concern-about-rule-of-zero.html

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+#include "SchemaTree.hpp"
+#include "SchemaTreeNode.hpp"
+#include "Value.hpp"
+
+namespace clp::ffi {
+/**
+ * A log event containing key-value pairs. Each event contains:
+ *  - A collection of node-id & value pairs, where each pair represents a leaf `SchemaTreeNode` in
+ *    the `SchemaTree`.
+ *   `SchemaTree` and the key is the node's ID.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:21]:
```suggestion
 * - A collection of node-ID & value pairs, where each pair represents a leaf `SchemaTreeNode` in
 *   the `SchemaTree`.
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.hpp [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+     * @param schema_tree
+     * @param node_id_value_pairs
+     * @param utc_offset
+     * @return A result containing the key-value pair log event or an error code indicating the
+     * failure:
+     * - std::errc::operation_not_permitted if a node ID doesn't represent a valid node in the
+     *   schema tree, or a non-leaf node ID is paired with a value.
+     * - std::errc::protocol_error if the schema tree node type doesn't match the value's type.
+     * - std::errc::protocol_not_supported if the same key appears more than once under a parent
+     *   node.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.hpp:41]:
```suggestion
     * failure. See `valdiate_node_id_value_pairs` for the possible error codes.
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+/**
+ * @param type
+ * @param value
+ * @return Whether the given schema tree node type matches the given value's type.
+ */
+[[nodiscard]] auto
+node_type_matches_value_type(SchemaTreeNode::Type type, Value const& value) -> bool;
+
+/**
+ * Validates whether the given node ID value pairs are valid in terms of the key-value pair IR spec.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:36]:
```suggestion
 * Validates whether the given node-ID value pairs are leaf nodes in the `SchemaTree` forming a
 * sub-tree of their own.
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+        SchemaTree const& schema_tree,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> std::optional<std::errc>;
+
+/**
+ * @param schema_tree
+ * @param node_id
+ * @param node_id_value_pairs
+ * @return Whether the given node is a leaf node in the sub schema tree defined by
+ * `node_id_value_pairs`. A node is considered a leaf if none of its descendants appear in the
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:56]:
```suggestion
 * @return Whether the given node is a leaf node in the sub-tree of the `SchemaTree` defined by
 * `node_id_value_pairs`. A node is considered a leaf if none of its descendants appear in
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+            return false;
+    }
+}
+
+auto validate_node_id_value_pairs(
+        SchemaTree const& schema_tree,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> std::optional<std::errc> {
+    std::optional<std::errc> ret_val;
+    std::unordered_map<SchemaTreeNode::id_t, std::unordered_set<std::string_view>>
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:90]: Let's move this into the try to reduce its scope.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+            } else if (false == node_type_matches_value_type(node_type, value.value())) {
+                ret_val.emplace(std::errc::protocol_error);
+                break;
+            }
+
+            if (SchemaTreeNode::Type::Obj == node_type
+                && false == is_leaf_node(schema_tree, node_id, node_id_value_pairs))
+            {
+                // Implicit key conflict: A `null` or empty value is given but its descendants
+                // appear in the node ID value pairs
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:116]:
```suggestion
                // The node's value is `null` or `{}` but its descendants appear in
                // `node_id_value_pairs`.
```

I think saying implicit/explicit key conflict is a bit more confusing than not.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+                ret_val.emplace(std::errc::protocol_not_supported);
+                break;
+            }
+
+            auto const parent_node_id{node.get_parent_id()};
+            auto const key_name{node.get_key_name()};
+            if (parent_node_id_to_key_names.contains(parent_node_id)) {
+                if (parent_node_id_to_key_names.at(parent_node_id).contains(key_name)) {
+                    // Explicit key conflict: the key is duplicated under the same parent
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:125]:
```suggestion
                    // The key is duplicated under the same parent
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    return ret_val;
+}
+
+auto is_leaf_node(
+        SchemaTree const& schema_tree,
+        SchemaTreeNode::id_t node_id,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> bool {
+    std::vector<SchemaTreeNode::id_t> dfs_stack;
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:144]: Have you considered using a `deque`?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+[[nodiscard]] auto
+node_type_matches_value_type(SchemaTreeNode::Type type, Value const& value) -> bool;
+
+/**
+ * Validates whether the given node ID value pairs are valid in terms of the key-value pair IR spec.
+ * @param schema_tree
+ * @param node_id_value_pairs
+ * @return `std::nullopt` if the inputs are valid, or an error code indicating the failure:
+ * - std::errc::operation_not_permitted if a node ID doesn't represent a valid node in the
+ *   schema tree, or a non-leaf node ID is paired with a value.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:41]: A non-leaf node paired with a value actually returns `protocol_not_supported`. That said, I think it should return `operation_not_permitted` since if it was permitted, `node_id_value_pairs` wouldn't be a tree.

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 28]
```diff
@@ -0,0 +1,529 @@
...
+ * @tparam Type The type to query.
+ * @param value The value to test against.
+ * @param typed_value The typed value to compare with.
+ */
+template <typename Type>
+auto test_value_get_immutable_view(Value const& value, Type const& typed_value) -> void;
+
+/**
+ * Generates invalid node ID value pairs with values that don't match type of the given schema tree
+ * node.
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:75]:
```suggestion
 * Generates invalid node-ID value pairs with values that don't match the type of the schema tree
 * node with the given ID.
```

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 29]
```diff
@@ -0,0 +1,529 @@
...
+            );
+        }
+    }
+    if (SchemaTreeNode::Type::Obj != node_type) {
+        invalid_node_id_value_pairs.emplace(node_id, std::nullopt);
+        invalid_node_id_value_pairs.emplace(node_id, Value{});
+    }
+}
+
+auto clone_value(Value const& value) -> Value {
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:219]: Hm, I missed this in the previous PRs but it seems like we should just have copy constructors for `Value` and `EncodedTextAst`. It feels like an overoptimization to force people to never use copy construction unless they write a method manually like this.

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 30]
```diff
@@ -0,0 +1,529 @@
...
+            {4, "d", SchemaTreeNode::Type::Float},
+            {3, "e", SchemaTreeNode::Type::Obj},
+            {4, "f", SchemaTreeNode::Type::Obj}
+    };
+    for (auto const& locator : locators) {
+        REQUIRE_NOTHROW(schema_tree->insert_node(locator));
+    }
+
+    {
+        // Empty invalid_node_id_value_pairs
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:351]: Overzealous refactoring.

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 31]
```diff
@@ -0,0 +1,529 @@
...
+    };
+    test_value_is<FourByteEncodedTextAst>(four_byte_encoded_text_ast_value);
+    test_value_get_immutable_view<FourByteEncodedTextAst>(
+            four_byte_encoded_text_ast_value,
+            get_encoded_text_ast<four_byte_encoded_variable_t>(cStringToEncode)
+    );
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("ffi_KeyValuePairLogEvent_create", "[ffi]") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:314]: You can reduce the amount of code for this test by using `SECTION` and even nested `SECTION` macros. Recall that CMake runs every `SECTION` from the beginning of the test, so the schema tree, node_id_value_pairs would be reset for each outer `SECTION`.

E.g., instead of cloning `valid_node_id_value_pairs`, you could just create it outside a `SECTION` and use it inside different `SECTION`s.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp (hunk 29 above)
Comment (kirkrodrigues) [reply components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:219]: What's bad about copying a `Value`? Is it not similar to copying a `string`?

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 26 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:144]: In the next PR, you will see that I'm actually using `std::stack` (which is implemented using `deque`) for dfs traversal, which is because the stack element is more complicated and using a vector means I need to provide move/copy semantics.
However, in this scenario, the stack is simply an int. I think the guideline of using `std` container is you should always try to use `vector` unless you have a strong motivation to use other ones. In this particular function, using `vector` have better locality, fewer indirections, and less memory allocation. Therefore, I would prefer to stick to `vector` unless you have other concern 😄

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp (hunk 29 above)
Comment (LinZhihao-723) [reply components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:219]: Copying string and `EncodedTextAst` can be expensive in deserialization. With copy disabled, we are more confident that there's no copy overhead in deserialization.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp (hunk 26 above)
Comment (kirkrodrigues) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:144]: Sure, was just asking.

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp (hunk 31 above)
Comment (kirkrodrigues) [reply components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:314]: 1. But by using sections, you'd only run the things outside the section(s) from the beginning, right?
2. Testing the same instance of the schema tree across *all* of the tests seems a bit extra since it should be const in the class anyway (and if it wasn't const, you're still relying on finding a bug in the implementation with a set of test cases that isn't exhaustive). Imo, having less code in the test is more valuable than testing that corner case.

From my reading, you could do something like:

1. Create schema tree
2. Section (optional): Test empty id-value pairs
3. Section (optional): Test mismatched types
4. Section (optional): Test valid id-value pairs
   * Section: Test duplicate key conflicts
   * Section: Test implicit key conflicts
5. Section (optional): Test out of bound node ID

That would allow you to at least get rid of the `clone_node_id_value_pairs` function (and make the code a bit easier to read).

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp (hunk 29 above)
Comment (LinZhihao-723) [reply components/core/tests/test-ffi_KeyValuePairLogEvent.cpp]: Adding move constructor for `EncodedTextAst` and `Value`

File: components/core/src/clp/ffi/Value.hpp [hunk 32]
```diff
@@ -150,12 +150,10 @@ class Value {
     template <FundamentalPrimitiveValueType T>
     explicit Value(T value) : m_value{value} {}
 
-    // Disable copy constructor and assignment operator
-    Value(Value const&) = delete;
-    auto operator=(Value const&) -> Value& = delete;
-
-    // Default move constructor and assignment operator
+    // Default copy/move constructor and assignment operator
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/Value.hpp:153]: Since everything except the constructor is default, we don't need to say they're default explicitly, right?

File: components/core/src/clp/ir/EncodedTextAst.hpp [hunk 33]
```diff
@@ -26,12 +26,10 @@ class EncodedTextAst {
               m_dict_vars{std::move(dict_vars)},
               m_encoded_vars{std::move(encoded_vars)} {}
 
-    // Disable copy constructor and assignment operator
-    EncodedTextAst(EncodedTextAst const&) = delete;
-    auto operator=(EncodedTextAst const&) -> EncodedTextAst& = delete;
-
-    // Default move constructor and assignment operator
+    // Default copy/move constructor and assignment operator
```
Comment (kirkrodrigues) [inline components/core/src/clp/ir/EncodedTextAst.hpp:29]: Ditto about default.

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 34]
```diff
@@ -0,0 +1,465 @@
...
+            {4, "d", SchemaTreeNode::Type::Str},
+            {4, "d", SchemaTreeNode::Type::Float},
+            {3, "e", SchemaTreeNode::Type::Obj},
+            {4, "f", SchemaTreeNode::Type::Obj}
+    };
+    for (auto const& locator : locators) {
+        REQUIRE_NOTHROW(schema_tree->insert_node(locator));
+    }
+
+    SECTION("Test empty id-value pairs") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:285]:
```suggestion
    SECTION("Test empty ID-value pairs") {
```

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 35]
```diff
@@ -0,0 +1,465 @@
...
+                    std::move(node_id_value_pair_to_test),
+                    UtcOffset{0}
+            )};
+            REQUIRE(result.has_error());
+            auto const& err{result.error()};
+            REQUIRE((std::errc::protocol_error == err));
+        }
+    }
+
+    SECTION("Test valid id-value pairs") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:359]:
```suggestion
    SECTION("Test valid ID-value pairs") {
```

File: components/core/tests/test-ffi_KeyValuePairLogEvent.cpp [hunk 36]
```diff
@@ -0,0 +1,465 @@
...
+        valid_node_id_value_pairs.emplace(10, Value{});
+        valid_node_id_value_pairs.emplace(11, std::nullopt);
+        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        auto const result{
+                KeyValuePairLogEvent::create(schema_tree, valid_node_id_value_pairs, UtcOffset{0})
+        };
+        REQUIRE_FALSE(result.has_error());
+
+        SECTION("Test duplicated key conflict on node #3") {
+            auto conflict_pairs{valid_node_id_value_pairs};
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_KeyValuePairLogEvent.cpp:399]: I thought the idea of using nested sections (besides breaking the code up) is that you wouldn't need to duplicate `valid_node_id_value_pairs` since it would start with a fresh state in each section, no? (You'd obviously call it something other than `valid_node_id_value_pairs`, but still.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — DISMISSED** [review]

**Review by kirkrodrigues — APPROVED** [review]
